grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer2.java

```diff
@@ -0,0 +1,487 @@
+      delegate.shutdown();
+      delegate = null;
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    shutdownDelegate();
+    shutdownLbComm();
+    for (Subchannel subchannel : subchannels.values()) {
+      subchannel.shutdown();
+    }
+    subchannels = Collections.emptyMap();
+  }
+
+  private void handleGrpclbError(Status status) {
+    if (roundRobinList.isEmpty()) {
+      helper.updatePicker(new ErrorPicker(status));
+    }
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    if (delegate != null) {
+      delegate.handleNameResolutionError(error);
+    } else {
+      handleGrpclbError(error);
+    }
+  }
+
+  private class LbResponseObserver implements StreamObserver<LoadBalanceResponse> {
+    boolean dismissed;
+
+    @Override public void onNext(final LoadBalanceResponse response) {
+      helper.runSerialized(new Runnable() {
+          @Override
+          public void run() {
+            handleResponse(response);
+          }
+        });
+    }
+
+    private void handleResponse(LoadBalanceResponse response) {
+      if (dismissed) {
+        return;
+      }
+      logger.log(Level.FINE, "[{0}] Got an LB response: {1}", new Object[] {logId, response});
+      // TODO(zhangkun83): make use of initialResponse
+      // InitialLoadBalanceResponse initialResponse = response.getInitialResponse();
+      ServerList serverList = response.getServerList();
+      HashMap<EquivalentAddressGroup, Subchannel> newSubchannelMap =
+          new HashMap<EquivalentAddressGroup, Subchannel>();
+      List<EquivalentAddressGroup> newRoundRobinList = new ArrayList<EquivalentAddressGroup>();
+      // TODO(zhangkun83): honor expiration_interval
+      // Construct the new collections. Create new Subchannels when necessary.
+      for (Server server : serverList.getServersList()) {
+        if (server.getDropRequest()) {
+          newRoundRobinList.add(null);
+        } else {
+          InetSocketAddress address;
+          try {
+            address = new InetSocketAddress(
+                InetAddress.getByAddress(server.getIpAddress().toByteArray()), server.getPort());
+          } catch (UnknownHostException e) {
+            handleGrpclbError(Status.UNAVAILABLE.withCause(e));
+            continue;
+          }
+          EquivalentAddressGroup eag = new EquivalentAddressGroup(address);
+          // TODO(zhangkun83): save the LB token and insert it to the application RPCs' headers.
+          if (!newSubchannelMap.containsKey(eag)) {
+            Attributes subchannelAttrs = Attributes.newBuilder()
+                .set(STATE_INFO, new AtomicReference<ConnectivityStateInfo>(
+                        ConnectivityStateInfo.forNonError(IDLE)))
+                .build();
+            Subchannel subchannel = helper.createSubchannel(eag, subchannelAttrs);
+            subchannel.requestConnection();
+            newSubchannelMap.put(eag, subchannel);
+          }
+          newRoundRobinList.add(eag);
```

Comment:
Duplicate EAGs from the naming system are intentionally preserved. They are used to distribute load unevenly across backends, a.k.a. weighted round-robin.